frontend/ui/fts.lua

```diff
@@ -0,0 +1,259 @@
+    fts.boottime_or_realtime_coarse = fts.realtime_coarse
+end
+
+--[[-- Alias for `monotonic_coarse`.
+
+The assumption being anything that requires accurate timestamps expects a monotonic clock source.
+This is certainly true for KOReader's UI scheduling.
+]]
+fts.now = fts.monotonic_coarse
+
+--- Converts an fts time to a Lua (decimal) number (sec.usecs) (accurate to the ms, rounded to 4 decimal places)
+function fts.tonumber(time_fts)
+    -- Round to 4 decimal places
+    return math.floor(fts.toSec(time_fts) * 10000) / 10000
+end
+
+-- Converts an fts to seconds (with comma)
+function fts.toSec(time_fts)
+    return time_fts / FTS_PRECISION
+end
+
+--[[-- Converts a fts to a Lua (int) number (resolution: 1ms).
+
+(Mainly useful when computing a time lapse for benchmarking purposes).
+]]
+function fts.tomSec(time_fts)
+    return math.floor(time_fts / FTS_PRECISION * 1e3 + 0.5)
+end
+
+--- Converts an fts to a Lua (int) number (resolution: 1µs)
+function fts.touSec(time_fts)
+
+    return math.floor(time_fts / FTS_PRECISION * 1e6 + 0.5)
+end
+
+--- Converts a Lua number (sec.usecs) to an fts time
+function fts.fromnumber(seconds)
+    return math.floor(seconds * FTS_PRECISION)
+end
+
+--[[-- Compare a past *MONOTONIC* fts time to *now*, returning the elapsed time between the two. (sec.usecs variant)
+
+Returns a Lua (decimal) number (sec.usecs) (accurate to the ms, rounded to 4 decimal places) (i.e., fts.tonumber())
+]]
+function fts.getDuration(start_fts)
+   return fts.tonumber(fts.now() - start_fts)
+end
+
+--[[-- Compare a past *MONOTONIC* fts time object to *now*, returning the elapsed time between the two. (ms variant)
+
+Returns a Lua (int) number (resolution: 1ms)
+]]
+function fts.getDurationMs(start_fts)
+   return fts.tomSec(fts.now() - start_fts)
+end
+
+--[[-- Compare a past *MONOTONIC* fts time object to *now*, returning the elapsed time between the two. (µs variant)
+
+Returns a Lua (int) number (resolution: 1µs)
+]]
+function fts.getDurationUs(start_fts)
+   return fts.touSec(fts.tonumber(start_fts))
+end
+
+--- Ditto for one set to math.huge
+fts.huge = math.huge
+
+function fts.fromTv(tv)
+    return tv.sec * FTS_PRECISION + tv.usec
+end
+
+function fts.splitsus(time_fts)
+    if not time_fts then return 0, 0 end
+    local sec = math.floor(time_fts / FTS_PRECISION)
+    local usec = math.floor(time_fts - sec * FTS_PRECISION)
+    return sec, usec
+end
+
+function fts.fromSec(seconds)
+    return math.floor(seconds * FTS_PRECISION)
+end
+
+function fts.frommSec(usec)
+    return math.floor(usec * FTS_PRECISION / 1e3)
+end
+
+function fts.fromuSec(usec)
+    return math.floor(usec * FTS_PRECISION / 1e6)
+end
+
+function fts.toTv(time_fts)
+    local TimeVal = require("ui/timeval")
+    if not time_fts then return TimeVal.zero end
+    local sec, usec = fts.splitsus(time_fts)
+    return TimeVal:new{ sec = sec, usec = usec }
+end
+
+-- for debugging
+function fts.format_fts(time_fts)
```

Comment:
I was going to say that there's a metatable function for that, but part of the whole exercise's point is to avoid a table + metatable object in the first place, so never mind me ;).